src/core/DY.Core.Trace.Invariant.fst

```diff
@@ -111,6 +113,26 @@ let rec trace_invariant #invs tr =
 
 (*** Lemmas on the trace invariant ***)
 
+/// If the trace invariant holds on a trace,
+/// then it must hold on the trace's prefixes.
+
+val trace_invariant_before:
```

Comment:
I don't see this lemma used anywhere else in this PR. It's a fairly natural thing to prove, but are there concrete cases where it becomes useful? If so, can we usefully put a pattern on this? It feels like the sort of lemma that one doesn't really want to have to manually invoke.

Reply:
In MLS* I use it when proving the various hypothesis of `guarded_authentication_lemma`. Since it only seems useful in that scenario, I would rather keep it without SMT patterns, if we need it often later we can always add the SMT pattern afterward.

I will add a comment above `guarded_authentication_lemma` mentioning this.

Reply:
Done